src/main/scala/is/hail/io/LoadMatrix.scala

```diff
@@ -330,15 +353,35 @@ object LoadMatrix {
           rvb.endArray()
           rvb.endStruct()
           rv.setOffset(rvb.end())
+          row += 1
           rv
         }
       }
 
+    val orderedRVD = if (optKeyExpr.isEmpty) {
+      val rvb = new RegionValueBuilder(Region())
+      rvb.start(TArray(TStruct("pk" -> TInt64())))
+      rvb.startArray(partitionCounts.length - 2)
+      var c = 0L
+      partitionCounts.slice(1, partitionCounts.length - 1).foreach { i =>
```

Comment:
I think it would be helpful to pull the bits from `val rvb = ...` to `ranges = new UnsafeIndexedSeq(...)` into a separate method.

This method is already massive as it is. Reading it and understanding all the invariants has been a challenge for me.

Reply:
And ideally we would also not switch on the missingness of the `optKeyExpr` again here, but I don't see a clear path to a cleaner implementation. Worth a think if you have the time. Maybe all the `optKeyExpr` stuff handling stuff needs to get pulled out into one or a few methods. Anything that makes it easier to see, in one place, what happens when there is and is not a key expr.